runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -293,22 +293,29 @@ State waitUntilFinish(
 
   @Override
   public State cancel() throws IOException {
-    Job content = new Job();
-    content.setProjectId(projectId);
-    content.setId(jobId);
-    content.setRequestedState("JOB_STATE_CANCELLED");
-    try {
-      dataflowOptions.getDataflowClient().projects().jobs()
-          .update(projectId, jobId, content)
-          .execute();
-    } catch (IOException e) {
-      String errorMsg = String.format(
-          "Failed to cancel the job, please go to the Developers Console to cancel it manually: %s",
-          MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
-      LOG.warn(errorMsg);
-      throw new IOException(errorMsg, e);
+    State state = getState();
+    if (!state.isTerminal()) {
+      Job content = new Job();
+      content.setProjectId(projectId);
+      content.setId(jobId);
+      content.setRequestedState("JOB_STATE_CANCELLED");
+      try {
+        dataflowOptions.getDataflowClient().projects().jobs()
+            .update(projectId, jobId, content)
+            .execute();
+      } catch (IOException e) {
```

Comment:
Do you mean that look at the IOException from execute() calls? Got confused how it will help to stop the race condition since the race condition is already happened. Or the "failure reason" means other messages?